### PR TITLE
[#400] Plan details playbooks support

### LINKED ISSUE #400: Pre & Post Playbook Support - Plan Details Page Changes
This issue is for the changes needed to the Details page to support pre- and post-migration playbooks.

### PR DESCRIPTION
**Pull request by michaelkro**
fixes #400 

## Notes
1.  ~Will squash all the WIP commits after review (reason: see 2)~ Done
2.  ~One way to test this locally is to `reset --hard HEAD^^`.  That will undo a cleanup commit and the commit that removes the mocked data/API responses.  Using the `nightly` db (or any db you have that has in progress plans), go to `Plans in Progress` and click on any plan.~  Things to check
    * `Download Log` dropdowns
    * Success notifications after "downloading" a playbook log (pre or post)
    * Info Popovers~
3.  Will need help verifying playbook log download implementation
4.  Should be merged with or after https://github.com/ManageIQ/manageiq-content/pull/355

### Running Post-migration Playbook
fullscreen_7_9_18__5_31_pm
fullscreen_7_9_18__5_25_pm-2

### No Playbooks
fullscreen_7_9_18__5_27_pm
fullscreen_7_9_18__5_28_pm

### Running Pre-migration Playbook
fullscreen_7_9_18__5_28_pm
fullscreen_7_9_18__5_29_pm

### Playbook Log Download Success
fullscreen_7_15_18__2_47_pm

### Playbook Log Download Failure
fullscreen_7_15_18__3_04_pm

### CODE REVIEW

File: app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js [hunk 1]
```diff
@@ -424,15 +437,20 @@ class PlanRequestDetailList extends React.Component {
...
                     ) : (
-                      <a
-                        href="#"
-                        onClick={e => {
-                          e.preventDefault();
-                          downloadLogAction(task);
-                        }}
+                      <DropdownButton
+                        id={`${task.id}-${task.descriptionPrefix}_download_log_dropdown`}
+                        title="Download Log"
```
Comment (priley86) [inline app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js]: __('Download Log')

File: app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js [hunk 2]
```diff
@@ -424,15 +437,20 @@ class PlanRequestDetailList extends React.Component {
...
+                        pullRight
+                        onSelect={eventKey => this.onSelect(eventKey, task)}
                       >
-                        {__('Download Log')}
-                      </a>
+                        {task.options.showPreMigrationOption && (
+                          <MenuItem eventKey="preMigration">Pre-migration log</MenuItem>
```
Comment (priley86) [inline app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js]: more 18n "Pre-migration" / "Post-migration"

File: app/javascript/react/screens/App/Plan/PlanActions.js [hunk 3]
```diff
@@ -10,11 +10,85 @@ import {
...
+            payload: task.id
+          });
+          const playbookLogFileName = `${task.vmName}-${playbookScheduleType}.log`;
+          const file = new File([response.data.stdout], playbookLogFileName, { type: 'text/plain;charset=utf-8' });
+          saveAs(file);
+          const successMsg = sprintf(__('"%s" download successful'), playbookLogFileName);
+          dispatch({
+            type: V2V_NOTIFICATION_ADD,
+            message: successMsg,
+            notificationType: 'success',
```
Comment (priley86) [inline app/javascript/react/screens/App/Plan/PlanActions.js]: nit... later on - we can move `notificationType` to a constant...

File: app/javascript/react/screens/App/Plan/PlanReducer.js [hunk 4]
```diff
@@ -57,6 +67,20 @@ const processVMTasks = vmTasks => {
       options: {}
     };
 
+    const hasPlaybookService = task.options.playbooks;
+
+    if (hasPlaybookService) {
+      taskDetails.options.prePlaybookRunning =
+        task.options.playbooks.pre && task.options.playbooks.pre.job_state === 'active';
```
Comment (priley86) [inline app/javascript/react/screens/App/Plan/PlanReducer.js]: we use PLAN_JOB_STATE constants later on - not a blocker at all right now

File: app/javascript/react/screens/App/Plan/PlanActions.js [hunk 5]
```diff
@@ -10,11 +10,85 @@ import {
...
+            persistent: true,
+            actionEnabled: false
+          });
+        })
+        .catch(e => {
+          dispatch({
+            type: DOWNLOAD_LOG_COMPLETED,
+            payload: task.id
+          });
+          dispatch({
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Plan/PlanActions.js]: I don't believe we need to add this notification, as this is a catch on an `API.get` miq notification system should take over signaling the error should any arise...

File: app/javascript/react/screens/App/Plan/PlanReducer.js [hunk 6]
```diff
@@ -57,6 +67,20 @@ const processVMTasks = vmTasks => {
       options: {}
     };
 
+    const hasPlaybookService = task.options.playbooks;
+
+    if (hasPlaybookService) {
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Plan/PlanReducer.js]: super minor nit, but we don't use `hasPlaybookService` anywhere but here, why not save a line, just do ` task.options.playbooks` here?

File: app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js [hunk 7]
```diff
@@ -233,6 +235,41 @@ class PlanRequestDetailList extends React.Component {
...
+      downloadLogAction(task);
+    } else {
+      fetchOrchestrationStackAction(fetchOrchestrationStackUrl, eventKey, task);
+    }
+  };
+
+  overlayTriggerClick = task => {
+    if (task.options.playbooks) {
+      const playbookStatuses = task.options.playbooks;
+      let runningPlaybook;
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js]: another nit... any chance we can set this = null?  I know I know, undefined will also be evaluated to false on line 258.... 😬 (if not 'saul good 🤗)

File: app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js [hunk 8]
```diff
@@ -432,15 +482,20 @@ class PlanRequestDetailList extends React.Component {
                         <span id="downloadLogInProgress">{__('Download log in progress...')}</span>
                       </label>
                     ) : (
-                      <a
-                        href="#"
-                        onClick={e => {
-                          e.preventDefault();
-                          downloadLogAction(task);
-                        }}
+                      <DropdownButton
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js]: Ok quick question.. I have no idea how this worked before... but If a user doesn't use a pre/post migration playbook, will they still have a log to download here? If so... maybe there should be a third option shown... to download that...

(before we had playbook support there were logs to download, just want to make sure we don't lose those)

Comment (AllenBW) [reply app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js]: ahh see thats what I meant! 😏 🤣

**Comment by AparnaKarve** [comment]
Looks good!

Will be testing this some more on the appliance today.

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 6d8758f2124b69374c256b42ab60187951a0966b
Author: Aparna Karve <aparna.karve@gmail.com>
Date:   Tue Jul 24 12:35:08 2018 -0700

    Merge pull request #442 from michaelkro/playbooks-plan-details
    
    [#400] Plan details playbooks support
    (cherry picked from commit 18e320d73da6da62b5ccf316d1ca0bb666c02790)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1608420
```

**Comment by simaishi** [comment]
@michaelkro @AparnaKarve Travis is now failing in Gaprindashvili branch. I think it's caused by this PR. Can you please take a look?

```
/home/travis/build/ManageIQ/miq_v2v_ui_plugin/app/javascript/react/screens/App/Plan/components/PlanRequestDetailList/PlanRequestDetailList.js
  212:3  error  renderInput should be placed after overlayTriggerClick  react/sort-comp
  238:3  error  onSelect should be placed before setPage                react/sort-comp
```
https://travis-ci.org/ManageIQ/miq_v2v_ui_plugin/jobs/410122179

**Comment by michaelkro** [comment]
@simaishi I've put up a PR against G to fix the errors https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/524

**Pull request merged**